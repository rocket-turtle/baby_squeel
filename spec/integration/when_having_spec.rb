require 'spec_helper'

describe '#when_having' do
  it 'adds a having clause' do
    relation = Post.selecting { id.count }
                   .grouping { author_id }
                   .when_having { id.count > 5 }

    expect(relation).to match_sql_snapshot
  end

  it 'adds a having clause with a calculation' do
    relation = Post.selecting { id.count }
                   .grouping { (author_id + 5 ) * 3 }
                   .when_having { id.count > 5 }

    expect(relation).to match_sql_snapshot
  end

  it 'adds a having clause with an association' do
    relation = Post.selecting { id.count }
                   .joins(:author)
                   .grouping { author.id }
                   .when_having { author.id.count > 5 }

    expect(relation).to match_sql_snapshot
  end

  it 'adds a having clause with an aliased table' do
    relation = Post.selecting { author.posts.id.count }
                   .joining { author.posts }
                   .grouping { author.posts.id }
                   .when_having { author.posts.id.count > 5 }

    expect(relation).to match_sql_snapshot(variants: ['8.1'])
  end
end
