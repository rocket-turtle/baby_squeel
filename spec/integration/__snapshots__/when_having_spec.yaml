---
"#when_having adds a having clause 1": SELECT COUNT("posts"."id") FROM "posts" GROUP
  BY "posts"."author_id" HAVING COUNT("posts"."id") > 5
"#when_having adds a having clause with a calculation 1": SELECT COUNT("posts"."id")
  FROM "posts" GROUP BY ("posts"."author_id" + 5) * 3 HAVING COUNT("posts"."id") >
  5
"#when_having adds a having clause with an association 1": SELECT COUNT("posts"."id")
  FROM "posts" INNER JOIN "authors" ON "authors"."id" = "posts"."author_id" GROUP
  BY "authors"."id" HAVING COUNT("authors"."id") > 5
"#when_having adds a having clause with an aliased table 1": SELECT COUNT("posts_authors"."id")
  FROM "posts" INNER JOIN "authors" ON "authors"."id" = "posts"."author_id" INNER
  JOIN "posts" "posts_authors" ON "posts_authors"."author_id" = "authors"."id" GROUP
  BY "posts_authors"."id" HAVING COUNT("posts_authors"."id") > 5
"#when_having adds a having clause with an aliased table 1 (Active Record: v8.1)": SELECT
  COUNT("posts_authors"."id") FROM "posts" INNER JOIN "authors" ON "authors"."id"
  = "posts"."author_id" INNER JOIN "posts" AS "posts_authors" ON "posts_authors"."author_id"
  = "authors"."id" GROUP BY "posts_authors"."id" HAVING COUNT("posts_authors"."id")
  > 5
